torch/distributed/_tensor/_dispatch.py

```diff
@@ -438,16 +441,20 @@ def _try_replicate_dtensor_spec_in_missing_dim(
         mesh: "DeviceMesh",
     ) -> DTensorSpec:
         # util function to produce a new spec for a DTensor arg/kwarg
-        # that puts Replicate() placement in the missing dimension for foreach ops
+        # that puts Replicate() placement in the missing dimension for foreach ops in 2D scenario
+        cur_mesh = dtensor_arg.device_mesh
+        error_msg = (
+            f"{op_call}: DTensor does not support cross-mesh operation yet! "
+            f"Got meshes: {mesh} {cur_mesh}"
+        )
+        # quick throw if implicit replication is not allowed
+        if not self._allow_implicit_replication:
+            raise NotImplementedError(error_msg)
+
         from torch.distributed.device_mesh import _mesh_resources
 
-        cur_mesh = dtensor_arg.device_mesh
         root_mesh = _mesh_resources.get_root_mesh(cur_mesh)
-        if (
-            self._allow_implicit_replication
-            and "foreach" in op_call.__name__
-            and root_mesh == mesh
-        ):
+        if "foreach" in op_call.__name__ and root_mesh == mesh:
```

Comment:
may I ask why we need `root_mesh == mesh`? what if `root_mesh` is 2D, but `mesh` is 1D.

Reply:
> may I ask why we need `root_mesh == mesh`? what if `root_mesh` is 2D, but `mesh` is 1D.

root_mesh should always == mesh after we record the largest mesh when implicit replication is turned on. This is to make sure the 1D DTensor has the same root_mesh as the other 2D DTensor in the foreach list.

Reply:
1. what if all DTensors are on a 2D submesh / 1D submesh of the 2D submesh, but the root_mesh is 3D. Should `foreach` still work in this case?
2. If the above case is impossible, can you show a concrete case when this check is useful?

Reply:
> 1. what if all DTensors are on a 2D submesh / 1D submesh of the 2D submesh, but the root_mesh is 3D. Should `foreach` still work in this case?
> 2. If the above case is impossible, can you show a concrete case when this check is useful?

Ah. I missed the 3D scenario. Ye. You are right. For 1, if all DTensors are on 2D/1D submesh of the 2D mesh, but the root mesh is 3D, it should continue to work. I modify the check to be checking whether the 1D and 2D mesh has the same root mesh instead.